working/0546-patterns/records-feature-specification.md

```diff
@@ -283,6 +283,41 @@ fields are) and collection literals.
 
 **TODO: Specify this more precisely.**
 
+### Constants
+
+A record expression is a compile-time constant expression
+if all its record field expressions are compile-time constant expressions. 
+
+_This is true whether the expression occurs in a constant context or not,
+which means that a record expression can be used directly as a parameter default value 
+if its record field expressions are constant expressions.
+Example: `f({(int, int) x = (1, 2)}) => ...`._
+
+A record expression is a potentially constant expression 
+if all its record field expressions are potentially constant or constant expressions.
+
+_This means that a record expression can be used in the initializer list
+of a constant non-redirecting generative constructor._
```

Comment:
> It means that we would allow declarations like `const C(int i): this.myRecord = (1, i);`.

That's actually a consequence of the "potentially constant" paragraph below, not this one. Its true that we'd allow it.

We should definitely not allow `const [1, i]` (with or without the leading `const`). We use different implementation classes with different behavior for constant and non-constant lists.
A `C(1, i)` (class constant constructor invocation with constant/potentially constant arguments) is more likely, because that uses the same class for constant and non-constant invocations, but because constructors can execute (some) code, we can get into problems with recursive constructor calls. And it differs wrt. canonicalization, which records don't (at least you can't tell if they do when a single record can be non-identical to itself).

Records are special. They may or may not correspond to an object at all, and you can't tell if they do.
A record like `(c1, c2)` really is just the juxtaposition of its values. It's just two values, nothing more or less. If each value can occur, so can both.

A list has structure on top of its contents. We use different classes for const lists and non-const lists. Creating a list is an object creation expression.

Not so for records. Example:
```dart
class C {
  final int a, b;
  final (int, int) p1;
  const C(int a, int b, int c, int d) : a = a, b = b, p = (c, d);
  (int, int) get p2 => (a, b);
}
```
There is no fundamental difference between the initializers for `a`, `b`, `p1.$0` and `p1.$1`, and I don't want there to be one. The only difference is the *grouping* of the values, where `p.$0` and `p.$1` can be accessed together as `p1`, just as `a` and `b` can be accessed through `p2`. The `p1` variable still just (potentially) two adjacent storage locations in the class.

We *can* add `const` in front of a record, but it feels spurious. It doesn't cause canonicalization, because records don't have a consistent identity anyway. It ensures that the field values are constants, but either that doesn't matter, it's enforced anyway, or you can put `const` on each field expression if necessary, like `({x = (const C(), const [])})`.
(I'd be fine with allowing `const expression` for any expression, which then puts that expression into a constant context. The `const (e1, e2)` would just be shorthand for `(const e1, const e2)`, and `const 4` is valid and constant.


I don't know if we intend to do the same thing for structs. If we *can*, I'd probably be happy with it. A `StructName(arg1,...,argn)` would be constant (aka. is valid where a constant is required) if it invokes the primary constructor, and the argument expressions are constant.
However, since a struct can have non-primary constructors too, and those would presumably require `const` to invoke, it would be better for consistency, with other struct constructors and other const constructors in general, to require a `const` in front. So, "probably not" to doing the same for structs, or other constructs with constructors.

Reply:
> Records are special. They may or may not correspond to an object at all, and you can't tell if they do.

I can't see any way to support a reference of type `Object` whose value is assigned/initialized to a record, other than boxing (or using an existing boxed representation of) the record, unless that whole reference can be removed during optimization. The type could also be any top type, or it could be a type variable whose value is `Object` or a top type.

This means that records can be known to have a heap entity representation at some point. Any usage of that representation at the record type might unbox it, but the `Object` typed reference is not a weak reference and the fact that it is a reference to a heap entity makes it more object-like than the wording

> It's just two values, nothing more or less

implies.

> We can add const in front of a record, but it feels spurious. It doesn't cause canonicalization

It _could_ cause canonicalization.

> allowing `const expression` for any expression

Looks like that would be a thin layer of syntactic sugar: Putting `const freshVariableName = expression;` at the top level and then replacing `const expression` by `freshVariableName`.